File: android/src/main/java/nz/co/electricbolt/appfigurateflutter/AppfigurateFlutterPlugin.java
package nz.co.electricbolt.appfigurateflutter;

import androidx.annotation.NonNull;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import nz.co.electricbolt.appfiguratelibrary.Appfigurate;
import nz.co.electricbolt.appfiguratelibrary.Configuration;
import nz.co.electricbolt.appfiguratelibrary.Log;

public class AppfigurateFlutterPlugin implements FlutterPlugin, MethodCallHandler, Appfigurate.ConfigurationUpdated {

  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "appfiguratelibrary");
    channel.setMethodCallHandler(this);
    Appfigurate.addConfigurationUpdatedListener(this);
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    Appfigurate.removeConfigurationUpdatedListener(this);
    channel.setMethodCallHandler(null);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("read")) {
      Log.log("Plugin: read");
      Map<String, Object> response = new HashMap<>();
      response.put("version", Appfigurate.version());
      response.put("description", Configuration.sharedConfiguration().description());
      response.put("modifications", Configuration.sharedConfiguration().modifications());
      response.put("configuration", Configuration.sharedConfiguration()._dictionaryFromConfiguration());
      result.success(response);
    } else if (call.method.equals("saveconfiguration")) {
      Log.log("Plugin: saveconfiguration");
      Appfigurate.saveConfiguration();
      result.success(null);
    } else if (call.method.equals("restoreconfiguration")) {
      Log.log("Plugin: restoreconfiguration");
      Appfigurate.restoreConfiguration();
      result.success(null);
    } else if (call.method.equals("setlogging")) {
      Log.log("Plugin: setlogging");
      Appfigurate.setLogging((Boolean) call.arguments);
      result.success(null);
    } else {
      Log.log("Plugin: flutter method " + call.method + " not implemented in Java");
      result.notImplemented();
    }
  }

  public void configurationUpdated(String action) {
    channel.invokeMethod("configurationupdated", action);
  }

}

File: ios/Classes/AppfigurateFlutterPlugin.m
// AppfigurateFlutterPlugin.m
// Appfigurate™ Copyright© 2022-2024; Electric Bolt Limited.

#import "AppfigurateFlutterPlugin.h"
@import AppfigurateLibrary;

@interface APLConfiguration ()

- (NSDictionary* _Nullable) dictionaryFromConfiguration;

@end

@interface AppfigurateFlutterPlugin () <APLConfigurationUpdated> {
    FlutterMethodChannel* channel;
}

@end;

@implementation AppfigurateFlutterPlugin

#pragma mark AppfigurateFlutterPlugin

+ (AppfigurateFlutterPlugin*) sharedInstance {
    static AppfigurateFlutterPlugin* instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [AppfigurateFlutterPlugin new];
    });
    return instance;
}

- (instancetype) init {
    self = [super init];
    if (self) {
    }
    return self;
}

+ (void) registerWithRegistrar: (nonnull NSObject<FlutterPluginRegistrar>*) registrar {
    [[AppfigurateFlutterPlugin sharedInstance] registerWithRegistrar: registrar];
}

- (void) registerWithRegistrar: (NSObject<FlutterPluginRegistrar>*) registrar {
    channel = [FlutterMethodChannel methodChannelWithName: @"appfiguratelibrary" binaryMessenger: [registrar messenger]];
    [registrar addMethodCallDelegate: [AppfigurateFlutterPlugin sharedInstance] channel: channel];
    [registrar addApplicationDelegate: self];
}

#pragma mark FlutterPlugin

- (BOOL) application: (UIApplication*) application didFinishLaunchingWithOptions: (NSDictionary*) launchOptions {
    APLDEBUG(@"Plugin: application:didFinishLaunchingWithOptions:");
    APLApplicationDidFinishLaunchingWithOptions(launchOptions);
    APLAddConfigurationUpdatedListener(self);
    return YES;
}

- (BOOL) application:(UIApplication*) application openURL: (NSURL*) url options: (NSDictionary<UIApplicationOpenURLOptionsKey, id>*) options {
    APLDEBUG(@"Plugin: application:openURL:options:");
    return APLApplicationOpenURL(url);
}

- (void) handleMethodCall: (FlutterMethodCall*) call result: (FlutterResult) result {
    if ([call.method isEqualToString: @"read"]) {
        APLDEBUG(@"Plugin: read");
        result(@{@"version": APLVersion(),
                 @"description": [[APLConfiguration sharedConfiguration] description],
                 @"modifications": [[APLConfiguration sharedConfiguration] modifications],
                 @"configuration": [[APLConfiguration sharedConfiguration] dictionaryFromConfiguration]
               });
    } else if ([call.method isEqualToString: @"saveconfiguration"]) {
        APLDEBUG(@"Plugin: saveconfiguration");
        APLSaveConfiguration();
        result(nil);
    } else if ([call.method isEqualToString: @"restoreconfiguration"]) {
        APLDEBUG(@"Plugin: restoreconfiguration");
        APLRestoreConfiguration();
        result(nil);
    } else if ([call.method isEqualToString: @"setlogging"]) {
        APLDEBUG(@"Plugin: setlogging");
        APLSetLogging((NSNumber*) call.arguments);
        result(nil);
    } else {
        APLDEBUG(@"Plugin: flutter method %@ not implemented in Objective-C", call.method);
        result(FlutterMethodNotImplemented);
    }
}

#pragma mark APLConfigurationUpdated

- (void) configurationUpdated: (NSNotification* _Nullable) notification {
    APLDEBUG(@"Plugin: configurationUpdated");
    [channel invokeMethod: @"configurationupdated" arguments: notification.userInfo[APLConfigurationUpdatedAction]];
}

@end

